crates/dbsp/src/operator/dynamic/balance/balancer.rs

```diff
@@ -0,0 +1,832 @@
+    /// policy (the idea being that the slowest worker will determine the overall speed), which
+    /// is in turn proportional to the amount of state maintained by the worker.
+    fn compute_domain(
+        &self,
+        metadata: &[Option<RebalancingExchangeSenderExchangeMetadata>],
+        hints: &BalancerHints,
+    ) -> BTreeMap<PartitioningPolicy, Cost> {
+        // Estimate the total size of the integral of the stream and the maximum size of a partition
+        // (assuming key-based partitioning).
+        //
+        // If a size hint is specified, ignore the actual key distribution and use the hint.
+        // If a skew hint is additionally specified, use it to compute the maximum size; otherwise
+        // assume even distribution.
+        //
+        // If no size hint is specified, compute max and total sizes from the actual key distribution.
+        let (total_size, max_size) = if let Some(size_hint) = hints.size_hint {
+            let max_size = if let Some(skew_hint) = hints.skew_hint {
+                ((size_hint as u64 / Runtime::num_workers() as u64) as f64 * skew_hint.into_inner())
+                    as u64
+            } else {
+                size_hint as u64 / Runtime::num_workers() as u64
+            };
+            (size_hint as u64, max_size)
+        } else {
+            let key_distribution =
+                RebalancingExchangeSenderExchangeMetadata::key_distribution(metadata);
+
+            let total_size = abs(key_distribution.total_records()) as u64;
+            let max_size = abs(key_distribution.sizes.iter().cloned().max().unwrap()) as u64;
+            (total_size, max_size)
+        };
+
+        BTreeMap::from([
+            // Shard: the cost is proportional to the maximum size of a partition.
+            (PartitioningPolicy::Shard, Cost(max_size)),
+            // Broadcast: every worker maintains the entire integral.
+            (PartitioningPolicy::Broadcast, Cost(total_size)),
+            // Balance: every worker maintains an equal share of the integral.
+            // the 1.5x factor discourages the use of this policy unless the skew is >1.5x.
+            (
+                PartitioningPolicy::Balance,
+                Cost(((total_size / Runtime::num_workers() as u64) as f64 * 1.5) as u64),
+            ),
+        ])
+    }
+
+    /// Set a hint for a stream.
+    ///
+    /// Fails if the hint specifies a fixed policy that cannot be enforced in the current state, i.e.,
+    /// it either contradicts the set of existing hints for other streams (e.g., we refuse to join
+    /// two streams both of which have Broadcast policies) or the policy cannot be enforced during the
+    /// current transaction because some of the streams can no longer be rebalanced.
+    ///
+    /// If the function succeeds, the stream will be successfully rebalanced using the specified policy.
+    fn set_hint(&mut self, node_id: NodeId, hint: BalancerHint) -> Result<(), BalancerError> {
```

Comment:
a problem with this API is that you may want to change several hints atomically, and no one order is legal, but changing them all would be